storage/replica_leader_lease.go

```diff
@@ -20,100 +20,164 @@
 package storage
 
 import (
+	"golang.org/x/net/context"
+
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/protoutil"
-	"golang.org/x/net/context"
 )
 
-// requestLeaderLease sends a request to obtain or extend a leader
-// lease for this replica. Unless an error is returned, the obtained
-// lease will be valid for a time interval containing the requested
-// timestamp. Only a single lease request may be pending at a time.
-func (r *Replica) requestLeaderLease(timestamp roachpb.Timestamp) <-chan *roachpb.Error {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+// pendingLeaderLeaseRequest coalesces LeaderLease requests and lets callers
+// join an in-progress one and wait for the result.
+// The actual execution of the LeaderLease Raft request is delegated to a
+// replica.
+// Methods are not thread-safe; a pendingLeaderLeaseRequest is logically part of
+// a replica, so replica.mu should be used to synchronize all calls.
+type pendingLeaderLeaseRequest struct {
+	// Slice of channels to send on after leader lease acquisition.
+	// If empty, then no request is in progress.
+	llChans []chan *roachpb.Error
+	// nextLease is the pending LeaderLease request, if any. It can be used to
+	// figure out if we're in the process of extending our own lease.
+	nextLease roachpb.Lease
+}
+
+// RequestPending returns the pending Lease, if one is in progress.
+// Otherwise, nil.
+func (p *pendingLeaderLeaseRequest) RequestPending() *roachpb.Lease {
+	pending := len(p.llChans) > 0
+	if pending {
+		return &p.nextLease
+	}
+	return nil
+}
 
+// InitOrJoinRequest executes a LeaderLease command asynchronously and returns a
+// promise for the result. If there's already a request in progress, we check
+// that the request is asking for a lease for the same replica as the one this
+// call wants to crown leader. If so, we join in waiting the results of that
+// request. If not, an error is immediately returned.
+//
+// replica is used to schedule and execute async work (proposing a LeaderLease
+// command). replica.mu is locked when delivering results, so calls from the
+// replica happen either before or after a result for a pending request has
+// happened.
+func (p *pendingLeaderLeaseRequest) InitOrJoinRequest(
+	replica *Replica,
+	nextLeader roachpb.ReplicaDescriptor,
+	timestamp roachpb.Timestamp,
+	startKey roachpb.Key,
+) <-chan *roachpb.Error {
 	llChan := make(chan *roachpb.Error, 1)
-	if len(r.mu.llChans) > 0 {
-		r.mu.llChans = append(r.mu.llChans, llChan)
+	if nextLease := p.RequestPending(); nextLease != nil {
+		if nextLease.Replica.ReplicaID == nextLeader.ReplicaID {
+			// Join a pending request asking for the same replica to become leader.
+			p.llChans = append(p.llChans, llChan)
+			return llChan
+		}
+		llChan <- roachpb.NewErrorf("request for different replica in progress "+
```

Comment:
This is a change in semantics and not a refactoring. Why is this in this commit? This will return errors all the way to the users if hit. Presumably we don't exercise this code yet, but the point remains.